pyvrp/cpp/educate/LocalSearch.cpp

```diff
@@ -172,6 +186,70 @@ bool LocalSearch::applyRouteOps(Route *U,
     return false;
 }
 
+void LocalSearch::maybeInsert(Node *U,
+                              Node *V,
+                              CostEvaluator const &costEvaluator)
+{
+    assert(!U->route && V->route);
+
+    auto const &uClient = data.client(U->client);
+
+    int const current = data.dist(V->client, n(V)->client);
+    int const proposed = data.dist(V->client, U->client)
+                         + data.dist(U->client, n(V)->client) - uClient.prize;
+
+    int deltaCost = proposed - current;
+
+    deltaCost += costEvaluator.loadPenalty(V->route->load() + uClient.demand,
+                                           data.vehicleCapacity());
+    deltaCost
+        -= costEvaluator.loadPenalty(V->route->load(), data.vehicleCapacity());
+
```

Comment:
We could add a precheck here
```suggestion
    if (deltaCost - V->route->timeWarp() >= 0)
        return deltaCost
```

Reply:
Maybe even
```suggestion
    if (deltaCost >= 0)
        return deltaCost
```
if we want to assume the timewarp can only increase by adding a customer, but this relies on subtle assumptions (like triangle inequality, see https://github.com/PyVRP/PyVRP/issues/56).

Reply:
I've opted for the second form. We probably won't lose much if we assume time warp does not go down, but I'd rather investigate that a bit later in a separate issue/PR. Strengthening these checks is probably worth it, but is probably a topic in and of itself.

Reply:
OK, no, I've tinkered with this a bit more and it does seem to be worthwhile doing the check involving time warp. I'm doing that, and will then run another benchmark to see if it helps.